File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/security/offline_mode/impl/WalletOfflineModeSettingsScreenImpl.java
package com.worldventures.dreamtrips.wallet.ui.settings.security.offline_mode.impl;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.jakewharton.rxbinding.widget.RxCompoundButton;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.core.utils.ProjectTextUtils;
import com.worldventures.dreamtrips.wallet.service.command.offline_mode.SwitchOfflineModeCommand;
import com.worldventures.dreamtrips.wallet.ui.common.base.WalletBaseController;
import com.worldventures.dreamtrips.wallet.ui.common.helper2.error.DialogErrorView;
import com.worldventures.dreamtrips.wallet.ui.common.helper2.error.ErrorViewFactory;
import com.worldventures.dreamtrips.wallet.ui.common.helper2.error.ErrorViewProvider;
import com.worldventures.dreamtrips.wallet.ui.common.helper2.progress.SimpleDialogProgressView;
import com.worldventures.dreamtrips.wallet.ui.settings.security.offline_mode.WalletOfflineModeSettingsPresenter;
import com.worldventures.dreamtrips.wallet.ui.settings.security.offline_mode.WalletOfflineModeSettingsScreen;
import com.worldventures.dreamtrips.wallet.ui.widget.WalletSwitcher;
import com.worldventures.dreamtrips.wallet.util.NetworkUnavailableException;

import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

import io.techery.janet.operationsubscriber.view.ComposableOperationView;
import io.techery.janet.operationsubscriber.view.ErrorView;
import io.techery.janet.operationsubscriber.view.OperationView;
import rx.Observable;

public class WalletOfflineModeSettingsScreenImpl extends WalletBaseController<WalletOfflineModeSettingsScreen, WalletOfflineModeSettingsPresenter> implements WalletOfflineModeSettingsScreen {

   private WalletSwitcher offlineModeSwitcher;

   @Inject WalletOfflineModeSettingsPresenter presenter;

   private Observable<Boolean> enableOfflineModeObservable;

   @Override
   protected void onFinishInflate(View view) {
      super.onFinishInflate(view);
      final Toolbar toolbar = view.findViewById(R.id.toolbar);
      toolbar.setNavigationOnClickListener(v -> getPresenter().goBack());
      final TextView tvPleaseNoteMessage = view.findViewById(R.id.offline_mode_please_note_label);
      tvPleaseNoteMessage.setText(ProjectTextUtils.fromHtml(getString(R.string.wallet_offline_mode_please_note_message)));
      offlineModeSwitcher = view.findViewById(R.id.offline_mode_switcher);
      enableOfflineModeObservable = RxCompoundButton.checkedChanges(offlineModeSwitcher).skip(1);
   }

   @Override
   public View inflateView(LayoutInflater layoutInflater, ViewGroup viewGroup) {
      return layoutInflater.inflate(R.layout.screen_wallet_settings_offline_mode, viewGroup, false);
   }

   @Override
   public boolean supportConnectionStatusLabel() {
      return false;
   }

   @Override
   public boolean supportHttpConnectionStatusLabel() {
      return false;
   }

   @Override
   public OperationView<SwitchOfflineModeCommand> provideOperationView() {
      return new ComposableOperationView<>(
            new SimpleDialogProgressView<>(getContext(),
                  ProjectTextUtils.fromHtml(getString(R.string.wallet_offline_mode_progress_message)), false),
            ErrorViewFactory.<SwitchOfflineModeCommand>builder()
                  .addProvider(getNetworkUnavailableDialogProvider())
                  .defaultErrorView(getDefaultErrorDialogProvider())
                  .build()
      );
   }

   @Override
   public Observable<Boolean> observeOfflineModeSwitcher() {
      return enableOfflineModeObservable;
   }

   @Override
   public void showConfirmationDialog(boolean enable) {
      new MaterialDialog.Builder(getContext())
            .content(enable ? R.string.wallet_offline_mode_enable_message : R.string.wallet_offline_mode_disable_message)
            .positiveText(R.string.wallet_continue_label)
            .onPositive((dialog, which) -> getPresenter().switchOfflineMode())
            .negativeText(R.string.cancel)
            .onNegative((dialog, which) -> getPresenter().switchOfflineModeCanceled())
            .cancelable(false)
            .build()
            .show();
   }

   @Override
   public void setOfflineModeState(boolean enabled) {
      offlineModeSwitcher.setCheckedWithoutNotify(enabled);
   }

   @NonNull
   private DialogErrorView<SwitchOfflineModeCommand> getDefaultErrorDialogProvider() {
      return new DialogErrorView<SwitchOfflineModeCommand>(getContext()) {
         @Override
         protected MaterialDialog createDialog(SwitchOfflineModeCommand command, Throwable throwable, Context context) {
            return new MaterialDialog.Builder(getContext())
                  .content(R.string.wallet_offline_mode_error_default_message)
                  .positiveText(R.string.retry)
                  .onPositive((dialog, which) -> getPresenter().switchOfflineMode())
                  .negativeText(R.string.cancel)
                  .onNegative((dialog, which) -> getPresenter().switchOfflineModeCanceled())
                  .cancelable(false)
                  .build();
         }
      };
   }

   @NonNull
   private ErrorViewProvider<SwitchOfflineModeCommand> getNetworkUnavailableDialogProvider() {
      return new ErrorViewProvider<SwitchOfflineModeCommand>() {
         @Override
         public Class<? extends Throwable> forThrowable() {
            return NetworkUnavailableException.class;
         }

         @Nullable
         @Override
         public ErrorView<SwitchOfflineModeCommand> create(SwitchOfflineModeCommand switchOfflineModeCommand, Throwable parentThrowable, Throwable throwable) {
            return new DialogErrorView<SwitchOfflineModeCommand>(getContext()) {
               @Override
               protected MaterialDialog createDialog(SwitchOfflineModeCommand command, Throwable throwable, Context context) {
                  return new MaterialDialog.Builder(getContext())
                        .content(R.string.wallet_offline_mode_error_no_internet_message)
                        .positiveText(R.string.settings)
                        .onPositive((dialog, which) -> getPresenter().navigateToSystemSettings())
                        .negativeText(R.string.cancel)
                        .onNegative((dialog, which) -> getPresenter().switchOfflineModeCanceled())
                        .cancelable(false)
                        .dismissListener(dialog -> getPresenter().fetchOfflineModeState())
                        .build();
               }
            };
         }
      };
   }

   @Override
   public WalletOfflineModeSettingsPresenter getPresenter() {
      return presenter;
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/security/offline_mode/impl/WalletOfflineModeSettingsScreenImpl.kt
package com.worldventures.wallet.ui.settings.security.offline_mode.impl

import android.content.Context
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.afollestad.materialdialogs.MaterialDialog
import com.jakewharton.rxbinding.widget.RxCompoundButton
import com.worldventures.core.utils.ProjectTextUtils
import com.worldventures.wallet.R
import com.worldventures.wallet.service.command.offline_mode.SwitchOfflineModeCommand
import com.worldventures.wallet.ui.common.base.WalletBaseController
import com.worldventures.wallet.ui.common.helper2.error.DialogErrorView
import com.worldventures.wallet.ui.common.helper2.error.ErrorViewFactory
import com.worldventures.wallet.ui.common.helper2.error.ErrorViewProvider
import com.worldventures.wallet.ui.common.helper2.progress.SimpleDialogProgressView
import com.worldventures.wallet.ui.settings.security.offline_mode.WalletOfflineModeSettingsPresenter
import com.worldventures.wallet.ui.settings.security.offline_mode.WalletOfflineModeSettingsScreen
import com.worldventures.wallet.ui.widget.WalletSwitcher
import com.worldventures.wallet.util.NetworkUnavailableException

import javax.inject.Inject

import io.techery.janet.operationsubscriber.view.ComposableOperationView
import io.techery.janet.operationsubscriber.view.ErrorView
import io.techery.janet.operationsubscriber.view.OperationView
import rx.Observable

class WalletOfflineModeSettingsScreenImpl : WalletBaseController<WalletOfflineModeSettingsScreen, WalletOfflineModeSettingsPresenter>(), WalletOfflineModeSettingsScreen {

   private lateinit var offlineModeSwitcher: WalletSwitcher
   @Inject lateinit var screenPresenter: WalletOfflineModeSettingsPresenter

   private lateinit var enableOfflineModeObservable: Observable<Boolean>

   private val defaultErrorDialogProvider: DialogErrorView<SwitchOfflineModeCommand>
      get() = object : DialogErrorView<SwitchOfflineModeCommand>(context) {
         override fun createDialog(command: SwitchOfflineModeCommand, throwable: Throwable, context: Context) =
               MaterialDialog.Builder(getContext())
                     .content(R.string.wallet_offline_mode_error_default_message)
                     .positiveText(R.string.wallet_retry_label)
                     .onPositive { _, _ -> presenter.switchOfflineMode() }
                     .negativeText(R.string.wallet_cancel_label)
                     .onNegative { _, _ -> presenter.switchOfflineModeCanceled() }
                     .cancelable(false)
                     .build()
      }

   private val networkUnavailableDialogProvider: ErrorViewProvider<SwitchOfflineModeCommand>
      get() = object : ErrorViewProvider<SwitchOfflineModeCommand> {
         override fun forThrowable() = NetworkUnavailableException::class.java

         override fun create(command: SwitchOfflineModeCommand, parentThrowable: Throwable?, throwable: Throwable):
               ErrorView<SwitchOfflineModeCommand> = object : DialogErrorView<SwitchOfflineModeCommand>(context) {
            override fun createDialog(command: SwitchOfflineModeCommand, throwable: Throwable, context: Context) =
                  MaterialDialog.Builder(getContext())
                        .content(R.string.wallet_offline_mode_error_no_internet_message)
                        .positiveText(R.string.wallet_settings)
                        .onPositive { _, _ -> presenter.navigateToSystemSettings() }
                        .negativeText(R.string.wallet_cancel_label)
                        .onNegative { _, _ -> presenter.switchOfflineModeCanceled() }
                        .cancelable(false)
                        .dismissListener { presenter.fetchOfflineModeState() }
                        .build()
         }
      }

   override fun onFinishInflate(view: View) {
      super.onFinishInflate(view)
      val toolbar = view.findViewById<Toolbar>(R.id.toolbar)
      toolbar.setNavigationOnClickListener { presenter.goBack() }
      val tvPleaseNoteMessage = view.findViewById<TextView>(R.id.offline_mode_please_note_label)
      tvPleaseNoteMessage.text = ProjectTextUtils.fromHtml(getString(R.string.wallet_offline_mode_please_note_message))
      offlineModeSwitcher = view.findViewById(R.id.offline_mode_switcher)
      enableOfflineModeObservable = RxCompoundButton.checkedChanges(offlineModeSwitcher).skip(1)
   }

   override fun inflateView(layoutInflater: LayoutInflater, viewGroup: ViewGroup): View =
         layoutInflater.inflate(R.layout.screen_wallet_settings_offline_mode, viewGroup, false)

   override fun supportConnectionStatusLabel() = false

   override fun supportHttpConnectionStatusLabel() = false

   override fun provideOperationView(): OperationView<SwitchOfflineModeCommand> {
      return ComposableOperationView(
            SimpleDialogProgressView(context,
                  ProjectTextUtils.fromHtml(getString(R.string.wallet_offline_mode_progress_message)), false),
            ErrorViewFactory.builder<SwitchOfflineModeCommand>()
                  .addProvider(networkUnavailableDialogProvider)
                  .defaultErrorView(defaultErrorDialogProvider)
                  .build()
      )
   }

   override fun observeOfflineModeSwitcher(): Observable<Boolean> = enableOfflineModeObservable

   override fun showConfirmationDialog(enable: Boolean) {
      MaterialDialog.Builder(context)
            .content(if (enable) R.string.wallet_offline_mode_enable_message else R.string.wallet_offline_mode_disable_message)
            .positiveText(R.string.wallet_continue_label)
            .onPositive { _, _ -> presenter.switchOfflineMode() }
            .negativeText(R.string.wallet_cancel_label)
            .onNegative { _, _ -> presenter.switchOfflineModeCanceled() }
            .cancelable(false)
            .build()
            .show()
   }

   override fun setOfflineModeState(enabled: Boolean) {
      offlineModeSwitcher.setCheckedWithoutNotify(enabled)
   }

   override fun getPresenter(): WalletOfflineModeSettingsPresenter = screenPresenter

   override fun screenModule() = WalletOfflineModeSettingsScreenModule()
}
